### PR TITLE
Feature evaluation and cluster interpretation for model

### LINKED ISSUE #194: Feature evaluation and cluster interpretation for GMM
Defining and evaluating features for GMM, interpreting clusters and making visualisations for report.

### PR DESCRIPTION
**Pull request by megan-pa**
## Type of Change
- [X] Feature
- [ ] Bugfix
- [ ] Both
- [ ] Other (please specify)

## What's been done
- Performed analysis on the 18 original features looking into their importance and relevance to training the model
- Plotted histograms trying to show that data follows a Gaussian distribution. Made plots for both 18 feature set and 8 feature set. 
- Identified and removed redundant features based on cluster mean separation and variance analysis 
- Changed features to be rate-based (actions per second) rather than count-based to remove the bias on session duration as this was causing issues with the clusters
- Reduced the feature set to 8 for more meaningful training 
- Trained and evaluated the GMM using BIC/AIC for model selection, and performed cross-validation for to find the best hyperparameters 
- Interpreted the clusters into different player behaviour types 
- Identified small issues with overlapping clusters and smaller cluster sizes (probably from noisy sessions and outliers)

## Screenshot(s) / Video(s)
Some examples of plots made. These will all be added to the report or at least in the appendix. First image is an example of one of the graphs trying to show the data is close enough to being normally distributed (I've done about six of these for comparison) and the other two are PCA visualisations of the clusters for 18 and 8 features. There are more plots and metrics than this though.

image

image

image

## How it's been done
- Preprocessing gameplay session data by filtering out shorter sessions and normalising features
- Engineered features so they are combined into rates and calculated accuracy metrics such as melee and spell accuracy 
- Trained the GMM across different combinations of clusters and covariance types
- Selected the final model hyperparameters using BIC primarily and used AIC as a more secondary metric
- Performed additional evaluation like PCA visualisation of clusters, cluster distance analysis and cluster size distribution

## What's next
- Will probably evaluate clusters again if we collect more data
- May refine the clusters and change the tactic model's pre-defined categories based on cluster interpretation
- Will eventually compare to the neural network approach I plan on implementing in the upcoming week
- Planning to add this evaluation to the report at some point
- Might move the evaluation code to a python notebook just for readability, but I think for now it's okay as is

## Related Issues
- #194 

## Notes for Reviewers
- Please let me know if you can think of any other evaluation I can perform on the model!

### CODE REVIEW

**Review by ethanr04 — COMMENTED** [review]
This all looks really great in terms of adding more tech and improving the ML model!

There is one slight issue: the second level is currently broken for me, and I can't move or anything. Hopefully this is a missing component or something and should be quick fix

**Comment by megan-pa** [comment]
> This all looks really great in terms of adding more tech and improving the ML model!
> 
> There is one slight issue: the second level is currently broken for me, and I can't move or anything. Hopefully this is a missing component or something and should be quick fix

Sorry that was an error from dev, there's a proper fix for it in my other PR but added a temporary fix in for now. Just need to make sure the hint is completed and it should work as normal.

**Review by ethanr04 — DISMISSED** [review]
Okay, that makes sense. After triggering the rewind hint it works as intended for me thanks!

**Review by qc23151 — COMMENTED** [review]
This branch works really well, would be happy to approve when merge conflicts are fixed. 😄

**Review by ethanr04 — APPROVED** [review]
All looking good post-merge

**Review by pabloapca — APPROVED** [review]
LGTM, good job!!
